### ISSUE TITLE
RHEL 9 support

### ISSUE THREAD

**Issue by Obihoernchen** [xcat2/xcat-core#7416]
This issue tracks the current status of RHEL 9 support in xCAT.

## Known issues

- [ ] nics.nicextraparams not working (still relying on ifcfg files)
- [ ] confignetwork not working with IPv6
- [ ] initscripts dependency missing. See: https://github.com/xcat2/xcat-core/issues/7343

Feel free to report additional issues.

**Comment by lbgracioso** [comment]
Hi @Obihoernchen 

I know this is about RHEL9 but I would like to add that RHEL9 based distros are also having issues.
`genimage` doesn't work in Rocky 9 (and I assume it doesn't work for OL and Alma either)

What I did to fix it:
I had to create a symbolic link for the following files and change their names to follow the naming pattern of each distribution in their respective folders (`/opt/xcat/share/xcat/netboot/<distro>`, `/opt/xcat/share/xcat/install/<distro>`).
```
/opt/xcat/share/xcat/netboot/rh/compute.rhels9.x86_64.pkglist
/opt/xcat/share/xcat/netboot/rh/compute.rhels9.x86_64.postinstall
/opt/xcat/share/xcat/netboot/rh/service.rhels9.x86_64.exlist
/opt/xcat/share/xcat/netboot/rh/service.rhels9.x86_64.otherpkgs.pkglist
/opt/xcat/share/xcat/netboot/rh/service.rhels9.x86_64.pkglist
/opt/xcat/share/xcat/netboot/rh/service.rhels9.x86_64.postinstall
/opt/xcat/share/xcat/install/rh/compute.rhels9.pkglist
/opt/xcat/share/xcat/install/rh/compute.rhels9.tmpl
/opt/xcat/share/xcat/install/rh/service.rhels9.pkglist
/opt/xcat/share/xcat/install/rh/service.rhels9.tmpl
/opt/xcat/share/xcat/install/rh/service.rhels9.x86_64.otherpkgs.pkglist
```

**Comment by Obihoernchen** [comment]
@lbgracioso that's fine. This issue is about all EL9 distros. I'll clarify this.

Good catch. I'll create the missing links.
Usually I use `rhels9` for all EL distros, that's why I didn't notice :D

**Comment by samveen** [comment]
@lbgracioso What you are suggesting is the exact implementation of the Centos7 provisioning support in xCAT (where the links were to the corresponding rhels7 templates/lists/{pre,post}scripts).

**Comment by ddj-brown** [comment]
Here are two examples that can be adopted 
-- bring up first IB interface with static address in a parallel subnet to primary
-- bring up separate IB partition from PKEY 1 (already exists in subnet manager) and get DHCP from master.
/install/postscripts/ipoib
------
#!/bin/bash

# Define the log function
function log {
	    echo "$(date +"%Y-%m-%d %H:%M:%S") - $1" >> /root/post.log 2>&1
	        logger -t xcat "$1"
	}

# Log script start
log "Starting ipoib script"

# Find the primary interface and its IPv4 address
primary_interface=$(ip route | awk '/default/ {print $5; exit}')
primary_ip=$(ip addr show dev $primary_interface | awk '$1 == "inet" {gsub(/\/.*$/, "", $2); print $2}')
log "Primary interface $primary_interface IPv4 address $primary_ip"

# Replace "172.20." with "172.25." in the primary IP address
ib_ip=${primary_ip/172.20./172.25.}
log "Using IP address $ib_ip for ipoib"

# Find first ib interface name with carrier up
ib_ifname=""
for (( i = 0; i < 5; i++ ))
do
	ib_ifname=$(ibdev2netdev | grep Up | head -1 | awk '{ print $5 }')
	[[ -n $ib_ifname ]] && break
	sleep 5
done

[[ -z $ib_ifname ]] && log "No IB device found with carrier"
[[ -n $ib_ifname ]] || exit
ib_conname=$ib_ifname

# Add InfiniBand connection
log "Adding InfiniBand connection $ib_ifname"
nmcli connection add type infiniband con-name $ib_conname ifname $ib_ifname transport-mode Datagram

# Set IPv4 address
log "Setting IPv4 address to $ib_ip/16"
nmcli connection modify $ib_conname ipv4.addresses "$ib_ip/16"

# Set IPv4 method to manual
log "Setting IPv4 method to manual"
nmcli connection modify $ib_conname ipv4.method manual

# Ignore IPv6
log "Ignoring IPv6"
nmcli connection modify $ib_conname ipv6.method ignore

# Activate the connection
log "Activating the connection"
nmcli connection up $ib_conname

# Log script end
log "/postscripts/ipoib script completed"
------
/install/postscripts/ibpkey -- bring up partition 1 with IP address from DHCP server
-----
#!/bin/bash

# Define the log function
function log {
	    echo "$(date +"%Y-%m-%d %H:%M:%S") - $1" >> /root/post.log 2>&1
	        logger -t xcat "$1"
	}

# Log script start
log "Starting ibpkey script"

# Find first ib interface name with carrier
# try five times with 5 second delay between attempts
ib_ifname=""
for (( i = 0; i < 5; i++ ))
do
        ib_ifname=$(ibdev2netdev | grep Up | head -1 | awk '{ print $5 }')
        [[ -n $ib_ifname ]] && break
        sleep 5
done
[[ -z $ib_ifname ]] && log "No IB device found with carrier"
[[ -n $ib_ifname ]] || exit
ib_devname=$(ibdev2netdev | grep $ib_ifname | awk '{ print $1 }')
ib_device=/sys/class/infiniband/$ib_devname/ports/1
if [[ -n $ib_device && -f $ib_device/lid ]]
then
	pkey=$(< $ib_device/pkeys/1)
	part=$(echo $pkey | sed 's/0x//')
fi
ib_conname=$ib_ifname.$part

# Add InfiniBand connection
log "Adding InfiniBand interface $ib_conname using interface $ib_ifname.$part"
nmcli connection add type infiniband con-name $ib_conname ifname $ib_ifname.$part
# Create child interface
log "Creating child interface $ib_conname of $ib_ifname for pkey $pkey"
nmcli connection modify $ib_conname infiniband.p-key $pkey infiniband.parent $ib_ifname

# Set IPv4 method to auto
log "Setting IPv4 method to auto"
nmcli connection modify $ib_conname ipv4.method auto

# Ignore IPv6
log "Ignoring IPv6"
nmcli connection modify $ib_conname ipv6.method ignore

# Activate the connection
log "Activating the connection"
nmcli connection up $ib_conname

# Log script end
log "/postscripts/ibpkey script completed"
-----

**Comment by ddj-brown** [comment]
[ibpostscripts.tar.gz](https://github.com/xcat2/xcat-core/files/14394244/ibpostscripts.tar.gz)
This contains two bash scripts that we're using on RHEL9.2
1) ipoib sets IP address for first IB interface with link.  Presumes boot net is 172.20.0.0/16 and IPoIB is 172.25.0.0/16 
2) ibpkey looks for a configured IB partition PKEY #1 and creates a subinterface, requesting IP from DHCP server on master node.

**Comment by samveen** [comment]
@Obihoernchen As per the github workflow cofiguration, the PR test checks the changes in the ubuntu 20.04 environment, which may not catch issues against RHEL and derivative environments. Can the PR testing be done in the RHEL and derivatives environment too, maybe Rocky 8 (for ease of access and use).

**Comment by Obihoernchen** [comment]
yes unfortunately :(
But the testing in the background is way bigger testing all different OSs already.
And also way more tests with real OS deployment etc.

**Comment by ZAM-** [comment]
Is this change only going to be for RHEL 9.x+ or earlier versions as well?

**Comment by samveen** [comment]
@ZAM- RHEL/Centos 7 family and RHEL/Centos/Rocky 8 family are already very well supported. This ticket is to track extending that support to include the RHEL9 and derivatives.

**Comment by rlcto** [comment]
@Obihoernchen Hi Markus, one question we have, and I think @ZAM-  was alluding to, is how we approach fixes related to networking.  By that I mean, a lot of the tools in xCAT were written to use the now deprecated ifcg commands and directory structure, e.g. ifcfg files vs. keyfiles.  My assumption is that we’re not replacing the ifcfg nomenclature but rather including the keyfile nomenclature.  I also assume that we would be doing a check to see if the host is RHEL9.  Do you have an example of how you have addressed this already if you have?

**Comment by Obihoernchen** [comment]
@rlcto @ZAM- Yes, often xCAT is using nmcli already, but in some parts (for instance nics.nicextraparams and I think IPv6) it relies on ifcfg files. In my opinion it would be best to use nmcli for everything and not rely on files at all.
To keep it simple, I guess current logic for EL<8 can stay as is, just add a special case for EL>=9 with full nmcli support if needed.
So far 90% is working with current code already (on EL9), just some minor parts are not using nmcli yet or use nmcli, but still rely on ifcfg files, too.
OS detection logic is included in networks scritps already. For instance: https://github.com/xcat2/xcat-core/blob/master/xCAT/postscripts/configeth#L706
Perhaps it is also easier to simply recreate the logic currently applied to the ifcfg files for the new keyfiles. But I would assume using nmcli should be easier.

**Comment by alexrichert** [comment]
@Obihoernchen in what you're envisioning as far as using nmcli for EL9, would the idea be that the user-provided settings would follow the current syntax/setting names/etc., or would the idea be that the user configs would directly specify nmcli arguments?

**Comment by Obihoernchen** [comment]
@alexrichert In my opinion everything in the nics table should stay as is. 90% of it is using nmcli already.
But there are some expections. For instance nics.nicextraparams relies on ifcfg files. For this nics column users should just be able to input any nmcli connection key-value pair into this field and the network scripts should just pass this 1:1 to nmcli I guess.
Furthermore even when using nmcli the scripts still write ifcfg files. This shouldn't happen on EL9+.

There might be other instances not using nmcli yet though. For now I only know about nics.nicextraparams.

**Comment by ocfmatt** [comment]
If you want to render something as a file or amend a file you could consider using keyfile format output before issuing a `nmcli con reload` command to make it active on a managed host. The use of a keyfile will help with extraparams logic but it will, as with ifcfg, rely on sane user input.

Network manager has been recommended by Red Hat since the introduction of RHEL8 and should, in my opinion, be the go to network management utility in EL8+ but I know some are familiar and reluctant to move away from ifcfg.

1. https://www.redhat.com/en/blog/rhel-9-networking-say-goodbye-ifcfg-files-and-hello-keyfiles
2. https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/assembly_networkmanager-connection-profiles-in-keyfile-format_configuring-and-managing-networking#proc_manually-creating-a-networkmanager-profile-in-keyfile-format_assembly_networkmanager-connection-profiles-in-keyfile-format

**Comment by LukeLR** [comment]
Is installing RHEL 9 already possible using xcat? Or do these issues still prevent an automated installation? We are still on CentOS 7 and want to switch to RockyLinux 9, but are waiting for xcat support. Thanks a lot!

**Comment by Obihoernchen** [comment]
Yes it's already possible with 2.16.5, but you might hit one of the confignetwork related issues mentioned above.
But overall stateless/stateful installation is working just fine.

**Comment by ddj-brown** [comment]
We have our cluster of 400+ nodes booted off rhel9 image from xcat server.  We replaced some of the postscripts to deal with NetworkManager, etc. There was a bit of a learning curve but nothing insurmountable. — setting static address on boot nic after getting dhcp— setting ipoib address based on boot nic address— adding a separate IB fabric key and requesting dhcp address on that subnet. I’m willing to share the scripts, let me know.   -- ddjDave ***@***.*** Jul 17, 2024, at 7:53 AM, L ***@***.***> wrote:﻿
Is installing RHEL 9 already possible using xcat? Or do these issues still prevent an automated installation? We are still on CentOS 7 and want to switch to RockyLinux 9, but are waiting for xcat support. Thanks a lot!

**Comment by mrobbert** [comment]
I would like to add a couple of tasks to this issue based on some problems that I am seeing with EL9 support.
The first is already an open issue from EL8, but it doesn't look like a fix ever got merged and I am seeing the same problem with Rocky 9. That problem is with adding OS update packages via additional pkgdir entries. The open issue for this problem is #7370 and it looks like there is a PR awaiting review, but it didn't get merged in 2.17 so I figured I'd mention it here to see if that can get merged for the next release.
The second issue is similar to the  issues that were recently fixed to improve EL9 support in the configeth and configib postscripts. We use the setroute postscript to modify the default gateway on our login nodes and that change is not being made persistent so it gets lost when a node reboots or even loses link on the external interface. This script calls the routeop script to make the changes and that script does not have any code that utilizes nmcli. If this problem needs its own issue please let me know and I can create that. Also let me know if you'd like me to make an attempt at a PR for this change. 

Thanks!

**Comment by Retzgesick** [comment]
> Is installing RHEL 9 already possible using xcat? Or do these issues still prevent an automated installation? We are still on CentOS 7 and want to switch to RockyLinux 9, but are waiting for xcat support. Thanks a lot!

>yes it's already possible with 2.16.5, but you might hit one of the confignetwork related issues mentioned above.
But overall stateless/stateful installation is working just fine.

Obviously several people got deploying RHEL9 working well. I am planning to upgrade an HPC cluster which is still on CentOS7.9 (!) to Rocky9.4. After rinstall, ssh to the newly deployed host asks for the password, but comes back then with a "Permission denied, please try again". I have to insert "PermitRootLogin yes" and restart sshd, then remote login with ssh succeeds. 

The content of /var/log/xcat/xcat.log is:
```
Running Kickstart Post-installation script...
Tue Jul  1 14:48:59 CEST 2025 [info]: xcat.deployment: Executing post.xcat to prepare for firstbooting ...
Tue Jul  1 14:49:18 CEST 2025 [info]: xcat.deployment: trying to download postscripts from 10.28.1.20...
Tue Jul  1 14:49:19 CEST 2025 [info]: xcat.deployment: postscripts downloaded successfully
Tue Jul  1 14:49:19 CEST 2025 [info]: xcat.deployment: trying to get mypostscript from 10.28.1.20...
Tue Jul  1 14:51:27 CEST 2025 [info]: xcat.deployment: finished firstboot preparation, sending request to 10.28.1.20:3002 for changing status...
ls: cannot access '/etc/sysconfig/network-scripts/ifcfg-*': No such file or directory
Tue Jul  1 02:56:31 PM CEST 2025 [error]: xcat.deployment: the network between the node and  is not ready, please check[retry=90]...
``` 
Between "and" and "is" are two spaces and normally there should be the ip of the xcat server (MASTER_IP). Why does the installation fail at this early point and does somebody have hints how I should proceed?

I am running xcat version 2.17

**Comment by pawsey-kbuckley** [comment]
**Rocky9 Kickstart file lists wrong repos in disable list**

We are running this version

```
# xcatsetup --version
Version 2.17.0 (git commit 2960b0e9f948abf4af8ce6c1f459191649883f15, built Wed Nov 13 15:34:26 CET 2024)
```
but I have also checked the files in a working copy of the repo

Noting this stanza from `./xCAT-server/share/xcat/install/scripts/post.rhels8`
(in which I have wrapped the list),
```
# List of internal repos to be disabled

internet_repo_file_list="oracle-linux-ol8.repo uek-ol8.repo \
 Rocky-AppStream.repo Rocky-BaseOS.repo Rocky-Extras.repo \
 CentOS-Base.repo \
 almalinux-ha.repo almalinux-nfv.repo almalinux-powertools.repo \
 almalinux.repo almalinux-resilientstorage.repo almalinux-rt.repo"

for repo_file in $internet_repo_file_list
do
  if [ -f /etc/yum.repos.d/$repo_file ]; then
    sed -i -e 's/enabled=1/enabled=0/' /etc/yum.repos.d/$repo_file
  fi
done
```

I find, that on a recently deployed Rocky 9.6 instance, these "repo files"
```
local-repository-0.repo:enabled=1		<== xCAT BaseOS
local-repository-0.repo:enabled=1		<== xCAT AppStream
rocky-extras.repo:enabled=1				<== rockylinux.org extras-versionnumber
rocky.repo:enabled=1					<== rockylinux.org BaseOS
rocky.repo:enabled=1					<== rockylinux.org AppStream
```

Note the case change in the names, and the fact that `BaseOS` and `AppStream`
are no longer listed in their own repo files.

That would suggest that, for RH9-based distros that (or at least Rocky9)
this change should be made to the in-line list

```
- Rocky-AppStream.repo Rocky-BaseOS.repo Rocky-Extras.repo
+ rocky.repo rocky-extras.repo
```
In my cloned copy of the xCAT repo, a `grep`  for `Rocky-AppStream.repo`
suggest that this file should be changed

`./xCAT-server/share/xcat/netboot/rh/genimage`

Then again, perhaps there needs to be a `post.rhels9` 

With thanks for xCAT,
Kevin Buckley

**Comment by viniciusferrao** [comment]
The networking and provisioning issues reported here have been addressed across several PRs:

**routeop/setroute not using nmcli** (raised by @mrobbert):
Fixed in #7530 — `routeop` was rewritten with full nmcli support. Routes are persisted via `nmcli connection modify` with `ipv4.routes`/`ipv6.routes` on EL9+, surviving reboots without ifcfg route files.

**ifcfg-\* missing during install on EL9** (raised by @Retzgesick):
The `post.rhels8` crash on `ls ifcfg-*` was fixed in #7530 — NM-only systems now use `nmcli con mod` to set autoconnect. The empty `MASTER_IP` in Retzgesick's report was a separate bug in `Template.pm` where `nodeset` silently rendered a kickstart with an empty master IP when hostname resolution failed. That's now fixed in #7545 (tracking issue: #7544) — `nodeset` will fail with a clear error instead of producing a broken kickstart.

**nicextraparams still writing ifcfg files on EL9** (raised by @Obihoernchen, @alexrichert):
Initially fixed in #7444 for `rhels9|alma9|rocky9`. Broadened in #7530 to match all EL9+ distro variants in both `configeth` and `configib`.

**configib not handling Rocky9 precreated NM connections** (raised by @alexrichert):
Fixed in #7459 and #7490.

**Rocky9 repo names in post scripts and genimage** (raised by @pawsey-kbuckley):
Fixed in #7530 — both `post.rhels8` and `genimage` now include `rocky.repo` and `rocky-extras.repo` alongside the old `Rocky-*.repo` names. Oracle Linux 9/10 repos also added.

**ospkgs/otherpkgs repo handling for EL9+**:
Fixed in #7530 — both scripts now use a shared `is_el_modular_pkgdir()` helper that covers all EL8+ variants including `almalinux`, which was previously missing.

**initscripts dependency / xcatd init path** (#7343):
Fix submitted in #7543 — replaces the SysV init script dependency with native systemd service management on EL7+ and Ubuntu 16.04+.

Once #7543 and #7545 merge, all items on this tracker will be resolved. @Obihoernchen thoughts on closing this after that?

**Comment by Obihoernchen** [comment]
Yes I agree. RHEL 9 support is in a very good state now.
Thank you very much!
